src/BatchReader/ClaimController.ts

```diff
@@ -1,30 +1,69 @@
 import { inject, injectable } from 'inversify'
-import { InsertWriteOpResult } from 'mongodb'
 
-import { DirectoryDAO } from './DirectoryDAO'
+import { asyncPipe } from 'Helpers/AsyncPipe'
+import { Database } from './Database'
 import { IPFS } from './IPFS'
+import { ReadEntry } from './ReadEntry'
+
+interface ReadFlowData {
+  readEntry?: ReadEntry
+  ipfsFileHashes?: ReadonlyArray<string>
+}
+
+type addEntries = (xs: ReadonlyArray<ReadEntry>) => Promise<void>
+
+type readFlow = (x?: ReadFlowData) => Promise<ReadFlowData>
 
 @injectable()
 export class ClaimController {
-  private readonly directoryDAO: DirectoryDAO
+  private readonly db: Database
   private readonly ipfs: IPFS
 
-  constructor(@inject('DirectoryDAO') directoryDAO: DirectoryDAO, @inject('IPFS') ipfs: IPFS) {
-    this.directoryDAO = directoryDAO
+  constructor(@inject('Database') db: Database, @inject('IPFS') ipfs: IPFS) {
+    this.db = db
     this.ipfs = ipfs
   }
 
-  addEntries = (entries: ReadonlyArray<{ ipfsDirectoryHash: string }>): Promise<InsertWriteOpResult> =>
-    this.directoryDAO.addEntries(entries)
-
-  readNextDirectory = async (): Promise<{ ipfsDirectoryHash: string; ipfsFileHashes: ReadonlyArray<string> }> => {
-    const collectionItem = await this.directoryDAO.findNextEntry()
-    if (!collectionItem) return
-    const { ipfsDirectoryHash } = collectionItem
-    await this.directoryDAO.incEntryAttempts({ ipfsDirectoryHash })
-    const ipfsFileHashes = await this.ipfs.getDirectoryFileHashes(ipfsDirectoryHash)
-    await this.directoryDAO.updateFileHashes({ ipfsDirectoryHash, ipfsFileHashes })
-    await this.directoryDAO.setEntrySuccessTime({ ipfsDirectoryHash })
-    return { ipfsDirectoryHash, ipfsFileHashes }
+  addEntries: addEntries = entries => this.db.readEntriesAdd(entries)
+
+  private readonly findNextEntry: readFlow = async () => {
+    const readEntry = await this.db.readEntryFindIncomplete()
+    return { readEntry }
+  }
+
+  private readonly verifyEntryWasFound: readFlow = async x =>
+    x.readEntry ? Promise.resolve(x) : Promise.reject('No entries remaining')
+
+  private readonly incEntryAttempts: readFlow = async ({ readEntry, ...rest }) => {
+    const updatedEntry = { ...readEntry, attempts: readEntry.attempts + 1 }
```

Comment:
Just realized this might be bad to increment attempts here due to race conditions. Handling this in mongodb would be better I think so we might need a specific function for this.